os/fs/smartfs/smartfs_utils.c

```diff
@@ -1268,6 +1268,127 @@ int smartfs_createentry(struct smartfs_mountpt_s *fs, uint16_t parentdirsector,
 	return ret;
 }
 
+/****************************************************************************
+ * Name: smartfs_replaceentry
+ *
+ * Description: Takes care of the operation of renaming an existing file
+ *              in a particular directory to a new name in the same dircetory
+ *              while minimizing the need to allocate new sectors for
+ *              writing the changed entry to MTD flash.
+ *
+ ****************************************************************************/
+
+int smartfs_replaceentry(struct smartfs_mountpt_s *fs, struct smartfs_entry_s oldentry, const char *filename, uint16_t type, struct smartfs_entry_s *direntry, FAR struct smartfs_ofile_s *sf)
+{
+	int ret;
+	struct smart_read_write_s readwrite;
+	uint16_t nextsector;
+	uint16_t entrysize;
+#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+	struct smartfs_chain_header_s *chainheader;
+#endif
+	uint8_t chainheader_type;
+
+	entrysize = sizeof(struct smartfs_entry_header_s) + fs->fs_llformat.namesize;
+
+	/*Read the entire logical sector containing the old entry into fs_rwbuffer*/
+	readwrite.logsector = oldentry.dsector;
+	readwrite.count = fs->fs_llformat.availbytes;
+	readwrite.offset = 0;
+	readwrite.buffer = (uint8_t *)fs->fs_rwbuffer;
+	ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
+	if (ret < 0) {
+		fdbg("Error %d reading sector %d data\n", ret, oldentry.dsector);
+		goto errout;
+	}
+
+	if (oldentry.firstsector == 0xFFFF) {
+		/* Allocate a new sector for the file / dir */
+
+		ret = FS_IOCTL(fs, BIOC_ALLOCSECT, 0xFFFF);
+		if (ret < 0) {
+			goto errout;
+		}
+
+		nextsector = (uint16_t)ret;
+
+		/* Set the newly allocated sector's type (file or dir) */
+
+#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+		if (sf) {
```

Comment:
Actually this is problem.
For open case we hold SF buffer.
This will update when it call write.
so we write entry in parent sector, so entry in parent sector points to invalidate sector(physically nothing written)
so before write, if power off happened, entry points empty sector. This will lead to read error.
so I added code to remove entry if child is not valid.. see smartfs_examine_sector, you can see code from below annotation.
```
/* If entry item is not valid, then remove from entry */
```
But actually this step bring some write operation, so if we also can hold this new entry sector in buffer, then everything is fine.
Please think about this carefully.